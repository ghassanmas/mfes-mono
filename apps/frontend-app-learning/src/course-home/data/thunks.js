import { logError } from '@edx/frontend-platform/logging';
import { camelCaseObject } from '@edx/frontend-platform';
import {
  executePostFromPostEvent,
  getCourseHomeCourseMetadata,
  getDatesTabData,
  getOutlineTabData,
  getProgressTabData,
  postCourseDeadlines,
  deprecatedPostCourseGoals,
  postWeeklyLearningGoal,
  postDismissWelcomeMessage,
  postRequestCert,
} from './api';

import {
  addModel,
} from '../../generic/model-store';

import {
  fetchTabDenied,
  fetchTabFailure,
  fetchTabRequest,
  fetchTabSuccess,
  setCallToActionToast,
} from './slice';

const eventTypes = {
  POST_EVENT: 'post_event',
};

export function fetchTab(courseId, tab, getTabData, targetUserId) {
  return async (dispatch) => {
    dispatch(fetchTabRequest({ courseId }));
    Promise.allSettled([
      getCourseHomeCourseMetadata(courseId, 'outline'),
      getTabData(courseId, targetUserId),
    ]).then(([courseHomeCourseMetadataResult, tabDataResult]) => {
      const fetchedCourseHomeCourseMetadata = courseHomeCourseMetadataResult.status === 'fulfilled';
      const fetchedTabData = tabDataResult.status === 'fulfilled';

      if (fetchedCourseHomeCourseMetadata) {
        dispatch(addModel({
          modelType: 'courseHomeMeta',
          model: {
            id: courseId,
            ...courseHomeCourseMetadataResult.value,
          },
        }));
      } else {
        logError(courseHomeCourseMetadataResult.reason);
      }

      if (fetchedTabData) {
        dispatch(addModel({
          modelType: tab,
          model: {
            id: courseId,
            ...tabDataResult.value,
          },
        }));
      } else {
        logError(tabDataResult.reason);
      }

      // Disable the access-denied path for now - it caused a regression
      if (fetchedCourseHomeCourseMetadata && !courseHomeCourseMetadataResult.value.courseAccess.hasAccess) {
        dispatch(fetchTabDenied({ courseId }));
      } else if (fetchedCourseHomeCourseMetadata && fetchedTabData) {
        dispatch(fetchTabSuccess({ courseId, targetUserId }));
      } else {
        dispatch(fetchTabFailure({ courseId }));
      }
    });
  };
}

export function fetchDatesTab(courseId) {
  return fetchTab(courseId, 'dates', getDatesTabData);
}

export function fetchProgressTab(courseId, targetUserId) {
  return fetchTab(courseId, 'progress', getProgressTabData, parseInt(targetUserId, 10) || targetUserId);
}

export function fetchOutlineTab(courseId) {
  return fetchTab(courseId, 'outline', getOutlineTabData);
}

export function dismissWelcomeMessage(courseId) {
  return async () => postDismissWelcomeMessage(courseId);
}

export function requestCert(courseId) {
  return async () => postRequestCert(courseId);
}

export function resetDeadlines(courseId, model, getTabData) {
  return async (dispatch) => {
    postCourseDeadlines(courseId, model).then(response => {
      const { data } = response;
      const {
        header,
        link,
        link_text: linkText,
      } = data;
      dispatch(getTabData(courseId));
      dispatch(setCallToActionToast({ header, link, linkText }));
    });
  };
}

export async function deprecatedSaveCourseGoal(courseId, goalKey) {
  return deprecatedPostCourseGoals(courseId, goalKey);
}

export async function saveWeeklyLearningGoal(courseId, daysPerWeek, subscribedToReminders) {
  return postWeeklyLearningGoal(courseId, daysPerWeek, subscribedToReminders);
}

export function processEvent(eventData, getTabData) {
  return async (dispatch) => {
    // Pulling this out early so the data doesn't get camelCased and is easier
    // to use when it's passed to the backend
    const { research_event_data: researchEventData } = eventData;
    const event = camelCaseObject(eventData);
    if (event.eventName === eventTypes.POST_EVENT) {
      executePostFromPostEvent(event.postData, researchEventData).then(response => {
        const { data } = response;
        const {
          header,
          link,
          link_text: linkText,
        } = data;
        dispatch(getTabData(event.postData.bodyParams.courseId));
        dispatch(setCallToActionToast({ header, link, linkText }));
      });
    }
  };
}
