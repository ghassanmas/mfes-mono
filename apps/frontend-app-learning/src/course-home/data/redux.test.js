import { Factory } from 'rosie';
import MockAdapter from 'axios-mock-adapter';

import { getAuthenticatedHttpClient } from '@edx/frontend-platform/auth';
import { getConfig } from '@edx/frontend-platform';

import * as thunks from './thunks';

import { appendBrowserTimezoneToUrl, executeThunk } from '../../utils';

import { initializeMockApp } from '../../setupTest';
import initializeStore from '../../store';

const { loggingService } = initializeMockApp();

const axiosMock = new MockAdapter(getAuthenticatedHttpClient());

describe('Data layer integration tests', () => {
  const courseHomeMetadata = Factory.build('courseHomeMetadata');
  const { id: courseId } = courseHomeMetadata;
  let courseMetadataUrl = `${getConfig().LMS_BASE_URL}/api/course_home/course_metadata/${courseId}`;
  courseMetadataUrl = appendBrowserTimezoneToUrl(courseMetadataUrl);

  let store;

  beforeEach(() => {
    axiosMock.reset();
    loggingService.logError.mockReset();

    store = initializeStore();
  });

  describe('Test fetchDatesTab', () => {
    const datesBaseUrl = `${getConfig().LMS_BASE_URL}/api/course_home/dates`;

    it('Should fail to fetch if error occurs', async () => {
      axiosMock.onGet(courseMetadataUrl).networkError();
      axiosMock.onGet(`${datesBaseUrl}/${courseId}`).networkError();

      await executeThunk(thunks.fetchDatesTab(courseId), store.dispatch);

      expect(loggingService.logError).toHaveBeenCalled();
      expect(store.getState().courseHome.courseStatus).toEqual('failed');
    });

    it('Should fetch, normalize, and save metadata', async () => {
      const datesTabData = Factory.build('datesTabData');

      const datesUrl = `${datesBaseUrl}/${courseId}`;

      axiosMock.onGet(courseMetadataUrl).reply(200, courseHomeMetadata);
      axiosMock.onGet(datesUrl).reply(200, datesTabData);

      await executeThunk(thunks.fetchDatesTab(courseId), store.dispatch);

      const state = store.getState();
      expect(state.courseHome.courseStatus).toEqual('loaded');
      expect(state).toMatchSnapshot();
    });
  });

  describe('Test fetchOutlineTab', () => {
    const outlineBaseUrl = `${getConfig().LMS_BASE_URL}/api/course_home/outline`;

    it('Should result in fetch failure if error occurs', async () => {
      axiosMock.onGet(courseMetadataUrl).networkError();
      axiosMock.onGet(`${outlineBaseUrl}/${courseId}`).networkError();

      await executeThunk(thunks.fetchOutlineTab(courseId), store.dispatch);

      expect(loggingService.logError).toHaveBeenCalled();
      expect(store.getState().courseHome.courseStatus).toEqual('failed');
    });

    it('Should fetch, normalize, and save metadata', async () => {
      const outlineTabData = Factory.build('outlineTabData', { courseId });

      const outlineUrl = `${outlineBaseUrl}/${courseId}`;

      axiosMock.onGet(courseMetadataUrl).reply(200, courseHomeMetadata);
      axiosMock.onGet(outlineUrl).reply(200, outlineTabData);

      await executeThunk(thunks.fetchOutlineTab(courseId), store.dispatch);

      const state = store.getState();
      expect(state.courseHome.courseStatus).toEqual('loaded');
      expect(state).toMatchSnapshot();
    });
  });

  describe('Test fetchProgressTab', () => {
    const progressBaseUrl = `${getConfig().LMS_BASE_URL}/api/course_home/progress`;

    it('Should result in fetch failure if error occurs', async () => {
      axiosMock.onGet(courseMetadataUrl).networkError();
      axiosMock.onGet(`${progressBaseUrl}/${courseId}`).networkError();

      await executeThunk(thunks.fetchProgressTab(courseId), store.dispatch);

      expect(loggingService.logError).toHaveBeenCalled();
      expect(store.getState().courseHome.courseStatus).toEqual('failed');
    });

    it('Should fetch, normalize, and save metadata', async () => {
      const progressTabData = Factory.build('progressTabData', { courseId });

      const progressUrl = `${progressBaseUrl}/${courseId}`;

      axiosMock.onGet(courseMetadataUrl).reply(200, courseHomeMetadata);
      axiosMock.onGet(progressUrl).reply(200, progressTabData);

      await executeThunk(thunks.fetchProgressTab(courseId), store.dispatch);

      const state = store.getState();
      expect(state.courseHome.courseStatus).toEqual('loaded');
      expect(state).toMatchSnapshot();
    });

    it('Should handle the url including a targetUserId', async () => {
      const progressTabData = Factory.build('progressTabData', { courseId });
      const targetUserId = 2;
      const progressUrl = `${progressBaseUrl}/${courseId}/${targetUserId}/`;

      axiosMock.onGet(courseMetadataUrl).reply(200, courseHomeMetadata);
      axiosMock.onGet(progressUrl).reply(200, progressTabData);

      await executeThunk(thunks.fetchProgressTab(courseId, 2), store.dispatch);

      const state = store.getState();
      expect(state.courseHome.targetUserId).toEqual(2);
    });
  });

  describe('Test saveCourseGoal', () => {
    it('Should save course goal', async () => {
      const goalUrl = `${getConfig().LMS_BASE_URL}/api/course_home/save_course_goal`;
      axiosMock.onPost(goalUrl).reply(200, {});

      await thunks.deprecatedSaveCourseGoal(courseId, 'unsure');

      expect(axiosMock.history.post[0].url).toEqual(goalUrl);
      expect(axiosMock.history.post[0].data).toEqual(`{"course_id":"${courseId}","goal_key":"unsure"}`);
    });
  });

  describe('Test resetDeadlines', () => {
    it('Should reset course deadlines', async () => {
      const resetUrl = `${getConfig().LMS_BASE_URL}/api/course_experience/v1/reset_course_deadlines`;
      const model = 'dates';
      axiosMock.onPost(resetUrl).reply(201, {});

      const getTabDataMock = jest.fn(() => ({
        type: 'MOCK_ACTION',
      }));

      await executeThunk(thunks.resetDeadlines(courseId, model, getTabDataMock), store.dispatch);

      expect(axiosMock.history.post[0].url).toEqual(resetUrl);
      expect(axiosMock.history.post[0].data).toEqual(`{"course_key":"${courseId}","research_event_data":{"location":"dates-tab"}}`);

      expect(getTabDataMock).toHaveBeenCalledWith(courseId);
    });
  });

  describe('Test dismissWelcomeMessage', () => {
    it('Should dismiss welcome message', async () => {
      const dismissUrl = `${getConfig().LMS_BASE_URL}/api/course_home/dismiss_welcome_message`;
      axiosMock.onPost(dismissUrl).reply(201);

      await executeThunk(thunks.dismissWelcomeMessage(courseId), store.dispatch);

      expect(axiosMock.history.post[0].url).toEqual(dismissUrl);
      expect(axiosMock.history.post[0].data).toEqual(`{"course_id":"${courseId}"}`);
    });
  });
});
