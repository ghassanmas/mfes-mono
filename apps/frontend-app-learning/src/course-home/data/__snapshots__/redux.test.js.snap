// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Data layer integration tests Test fetchDatesTab Should fetch, normalize, and save metadata 1`] = `
Object {
  "courseHome": Object {
    "courseId": "course-v1:edX+DemoX+Demo_Course",
    "courseStatus": "loaded",
    "proctoringPanelStatus": "loading",
    "targetUserId": undefined,
    "toastBodyLink": null,
    "toastBodyText": null,
    "toastHeader": "",
  },
  "courseware": Object {
    "courseId": null,
    "courseStatus": "loading",
    "sequenceId": null,
    "sequenceMightBeUnit": false,
    "sequenceStatus": "loading",
  },
  "models": Object {
    "courseHomeMeta": Object {
      "course-v1:edX+DemoX+Demo_Course": Object {
        "canLoadCourseware": true,
        "celebrations": null,
        "courseAccess": Object {
          "additionalContextUserMessage": null,
          "developerMessage": null,
          "errorCode": null,
          "hasAccess": true,
          "userFragment": null,
          "userMessage": null,
        },
        "id": "course-v1:edX+DemoX+Demo_Course",
        "isEnrolled": false,
        "isMasquerading": false,
        "isSelfPaced": false,
        "isStaff": false,
        "number": "DemoX",
        "org": "edX",
        "originalUserIsStaff": false,
        "start": "2013-02-05T05:00:00Z",
        "tabs": Array [
          Object {
            "slug": "outline",
            "title": "Course",
            "url": "http://localhost:18000/courses/course-v1:edX+DemoX+Demo_Course/course/",
          },
          Object {
            "slug": "discussion",
            "title": "Discussion",
            "url": "http://localhost:18000/courses/course-v1:edX+DemoX+Demo_Course/discussion/forum/",
          },
          Object {
            "slug": "wiki",
            "title": "Wiki",
            "url": "http://localhost:18000/courses/course-v1:edX+DemoX+Demo_Course/course_wiki",
          },
          Object {
            "slug": "progress",
            "title": "Progress",
            "url": "http://localhost:18000/courses/course-v1:edX+DemoX+Demo_Course/progress",
          },
          Object {
            "slug": "instructor",
            "title": "Instructor",
            "url": "http://localhost:18000/courses/course-v1:edX+DemoX+Demo_Course/instructor",
          },
          Object {
            "slug": "dates",
            "title": "Dates",
            "url": "http://localhost:18000/courses/course-v1:edX+DemoX+Demo_Course/dates",
          },
        ],
        "title": "Demonstration Course",
        "userTimezone": "UTC",
        "username": "MockUser",
        "verifiedMode": Object {
          "accessExpirationDate": null,
          "currency": "USD",
          "currencySymbol": "$",
          "price": 149,
          "sku": "8CF08E5",
          "upgradeUrl": "http://localhost:18130/basket/add/?sku=8CF08E5",
        },
      },
    },
    "dates": Object {
      "course-v1:edX+DemoX+Demo_Course": Object {
        "courseDateBlocks": Array [
          Object {
            "date": "2020-05-01T17:59:41Z",
            "dateType": "course-start-date",
            "description": "",
            "extraInfo": null,
            "learnerHasAccess": true,
            "link": "",
            "title": "Course Starts",
          },
          Object {
            "assignmentType": "Homework",
            "complete": true,
            "date": "2020-05-04T02:59:40.942669Z",
            "dateType": "assignment-due-date",
            "description": "",
            "extraInfo": null,
            "learnerHasAccess": true,
            "title": "Multi Badges Completed",
          },
          Object {
            "assignmentType": "Homework",
            "date": "2020-05-05T02:59:40.942669Z",
            "dateType": "assignment-due-date",
            "description": "",
            "extraInfo": null,
            "learnerHasAccess": true,
            "title": "Multi Badges Past Due",
          },
          Object {
            "assignmentType": "Homework",
            "date": "2020-05-27T02:59:40.942669Z",
            "dateType": "assignment-due-date",
            "description": "",
            "extraInfo": null,
            "learnerHasAccess": true,
            "link": "https://example.com/",
            "title": "Both Past Due 1",
          },
          Object {
            "assignmentType": "Homework",
            "date": "2020-05-27T02:59:40.942669Z",
            "dateType": "assignment-due-date",
            "description": "",
            "extraInfo": null,
            "learnerHasAccess": true,
            "link": "https://example.com/",
            "title": "Both Past Due 2",
          },
          Object {
            "assignmentType": "Homework",
            "complete": true,
            "date": "2020-05-28T08:59:40.942669Z",
            "dateType": "assignment-due-date",
            "description": "",
            "extraInfo": null,
            "learnerHasAccess": true,
            "link": "https://example.com/",
            "title": "One Completed/Due 1",
          },
          Object {
            "assignmentType": "Homework",
            "date": "2020-05-28T08:59:40.942669Z",
            "dateType": "assignment-due-date",
            "description": "",
            "extraInfo": null,
            "learnerHasAccess": true,
            "link": "https://example.com/",
            "title": "One Completed/Due 2",
          },
          Object {
            "assignmentType": "Homework",
            "complete": true,
            "date": "2020-05-29T08:59:40.942669Z",
            "dateType": "assignment-due-date",
            "description": "",
            "extraInfo": null,
            "learnerHasAccess": true,
            "link": "https://example.com/",
            "title": "Both Completed 1",
          },
          Object {
            "assignmentType": "Homework",
            "complete": true,
            "date": "2020-05-29T08:59:40.942669Z",
            "dateType": "assignment-due-date",
            "description": "",
            "extraInfo": null,
            "learnerHasAccess": true,
            "link": "https://example.com/",
            "title": "Both Completed 2",
          },
          Object {
            "date": "2020-06-16T17:59:40.942669Z",
            "dateType": "verified-upgrade-deadline",
            "description": "Don't miss the opportunity to highlight your new knowledge and skills by earning a verified certificate.",
            "extraInfo": null,
            "learnerHasAccess": true,
            "link": "https://example.com/",
            "title": "Upgrade to Verified Certificate",
          },
          Object {
            "assignmentType": "Homework",
            "date": "2030-08-17T05:59:40.942669Z",
            "dateType": "assignment-due-date",
            "description": "",
            "extraInfo": null,
            "learnerHasAccess": false,
            "link": "https://example.com/",
            "title": "One Verified 1",
          },
          Object {
            "assignmentType": "Homework",
            "date": "2030-08-17T05:59:40.942669Z",
            "dateType": "assignment-due-date",
            "description": "",
            "extraInfo": null,
            "learnerHasAccess": true,
            "link": "https://example.com/",
            "title": "One Verified 2",
          },
          Object {
            "assignmentType": "Homework",
            "date": "2030-08-17T05:59:40.942669Z",
            "dateType": "assignment-due-date",
            "description": "",
            "extraInfo": "ORA Dates are set by the instructor, and can't be changed",
            "learnerHasAccess": true,
            "link": "https://example.com/",
            "title": "ORA Verified 2",
          },
          Object {
            "assignmentType": "Homework",
            "date": "2030-08-18T05:59:40.942669Z",
            "dateType": "assignment-due-date",
            "description": "",
            "extraInfo": null,
            "learnerHasAccess": false,
            "link": "https://example.com/",
            "title": "Both Verified 1",
          },
          Object {
            "assignmentType": "Homework",
            "date": "2030-08-18T05:59:40.942669Z",
            "dateType": "assignment-due-date",
            "description": "",
            "extraInfo": null,
            "learnerHasAccess": false,
            "link": "https://example.com/",
            "title": "Both Verified 2",
          },
          Object {
            "assignmentType": "Homework",
            "date": "2030-08-19T05:59:40.942669Z",
            "dateType": "assignment-due-date",
            "description": "",
            "learnerHasAccess": true,
            "title": "One Unreleased 1",
          },
          Object {
            "assignmentType": "Homework",
            "date": "2030-08-19T05:59:40.942669Z",
            "dateType": "assignment-due-date",
            "description": "",
            "extraInfo": null,
            "learnerHasAccess": true,
            "link": "https://example.com/",
            "title": "One Unreleased 2",
          },
          Object {
            "assignmentType": "Homework",
            "date": "2030-08-20T05:59:40.942669Z",
            "dateType": "assignment-due-date",
            "description": "",
            "extraInfo": null,
            "learnerHasAccess": true,
            "title": "Both Unreleased 1",
          },
          Object {
            "assignmentType": "Homework",
            "date": "2030-08-20T05:59:40.942669Z",
            "dateType": "assignment-due-date",
            "description": "",
            "extraInfo": null,
            "learnerHasAccess": true,
            "title": "Both Unreleased 2",
          },
          Object {
            "date": "2030-08-23T00:00:00Z",
            "dateType": "course-end-date",
            "description": "",
            "extraInfo": null,
            "learnerHasAccess": true,
            "link": "",
            "title": "Course Ends",
          },
          Object {
            "date": "2030-09-01T00:00:00Z",
            "dateType": "verification-deadline-date",
            "description": "You must successfully complete verification before this date to qualify for a Verified Certificate.",
            "extraInfo": null,
            "learnerHasAccess": false,
            "link": "https://example.com/",
            "title": "Verification Deadline",
          },
        ],
        "datesBannerInfo": Object {
          "contentTypeGatingEnabled": false,
          "missedDeadlines": false,
          "missedGatedContent": false,
          "verifiedUpgradeLink": "http://localhost:18130/basket/add/?sku=8CF08E5",
        },
        "hasEnded": false,
        "id": "course-v1:edX+DemoX+Demo_Course",
        "learnerIsFullAccess": true,
      },
    },
  },
  "recommendations": Object {
    "recommendationsStatus": "loading",
  },
  "tours": Object {
    "showCoursewareTour": false,
    "showExistingUserCourseHomeTour": false,
    "showNewUserCourseHomeModal": false,
    "showNewUserCourseHomeTour": false,
    "toursEnabled": false,
  },
}
`;

exports[`Data layer integration tests Test fetchOutlineTab Should fetch, normalize, and save metadata 1`] = `
Object {
  "courseHome": Object {
    "courseId": "course-v1:edX+DemoX+Demo_Course",
    "courseStatus": "loaded",
    "proctoringPanelStatus": "loading",
    "targetUserId": undefined,
    "toastBodyLink": null,
    "toastBodyText": null,
    "toastHeader": "",
  },
  "courseware": Object {
    "courseId": null,
    "courseStatus": "loading",
    "sequenceId": null,
    "sequenceMightBeUnit": false,
    "sequenceStatus": "loading",
  },
  "models": Object {
    "courseHomeMeta": Object {
      "course-v1:edX+DemoX+Demo_Course": Object {
        "canLoadCourseware": true,
        "celebrations": null,
        "courseAccess": Object {
          "additionalContextUserMessage": null,
          "developerMessage": null,
          "errorCode": null,
          "hasAccess": true,
          "userFragment": null,
          "userMessage": null,
        },
        "id": "course-v1:edX+DemoX+Demo_Course",
        "isEnrolled": false,
        "isMasquerading": false,
        "isSelfPaced": false,
        "isStaff": false,
        "number": "DemoX",
        "org": "edX",
        "originalUserIsStaff": false,
        "start": "2013-02-05T05:00:00Z",
        "tabs": Array [
          Object {
            "slug": "outline",
            "title": "Course",
            "url": "http://localhost:18000/courses/course-v1:edX+DemoX+Demo_Course/course/",
          },
          Object {
            "slug": "discussion",
            "title": "Discussion",
            "url": "http://localhost:18000/courses/course-v1:edX+DemoX+Demo_Course/discussion/forum/",
          },
          Object {
            "slug": "wiki",
            "title": "Wiki",
            "url": "http://localhost:18000/courses/course-v1:edX+DemoX+Demo_Course/course_wiki",
          },
          Object {
            "slug": "progress",
            "title": "Progress",
            "url": "http://localhost:18000/courses/course-v1:edX+DemoX+Demo_Course/progress",
          },
          Object {
            "slug": "instructor",
            "title": "Instructor",
            "url": "http://localhost:18000/courses/course-v1:edX+DemoX+Demo_Course/instructor",
          },
          Object {
            "slug": "dates",
            "title": "Dates",
            "url": "http://localhost:18000/courses/course-v1:edX+DemoX+Demo_Course/dates",
          },
        ],
        "title": "Demonstration Course",
        "userTimezone": "UTC",
        "username": "MockUser",
        "verifiedMode": Object {
          "accessExpirationDate": null,
          "currency": "USD",
          "currencySymbol": "$",
          "price": 149,
          "sku": "8CF08E5",
          "upgradeUrl": "http://localhost:18130/basket/add/?sku=8CF08E5",
        },
      },
    },
    "outline": Object {
      "course-v1:edX+DemoX+Demo_Course": Object {
        "accessExpiration": null,
        "canShowUpgradeSock": false,
        "certData": Object {
          "certStatus": null,
          "certWebViewUrl": null,
          "certificateAvailableDate": null,
          "downloadUrl": null,
        },
        "courseBlocks": Object {
          "courses": Object {
            "block-v1:edX+DemoX+Demo_Course+type@course+block@bcdabcdabcdabcdabcdabcdabcdabcd3": Object {
              "hasScheduledContent": false,
              "id": "course-v1:edX+DemoX+Demo_Course",
              "sectionIds": Array [
                "block-v1:edX+DemoX+Demo_Course+type@chapter+block@bcdabcdabcdabcdabcdabcdabcdabcd2",
              ],
              "title": "bcdabcdabcdabcdabcdabcdabcdabcd3",
            },
          },
          "sections": Object {
            "block-v1:edX+DemoX+Demo_Course+type@chapter+block@bcdabcdabcdabcdabcdabcdabcdabcd2": Object {
              "complete": false,
              "courseId": "course-v1:edX+DemoX+Demo_Course",
              "id": "block-v1:edX+DemoX+Demo_Course+type@chapter+block@bcdabcdabcdabcdabcdabcdabcdabcd2",
              "resumeBlock": false,
              "sequenceIds": Array [
                "block-v1:edX+DemoX+Demo_Course+type@sequential+block@bcdabcdabcdabcdabcdabcdabcdabcd1",
              ],
              "title": "Title of Section",
            },
          },
          "sequences": Object {
            "block-v1:edX+DemoX+Demo_Course+type@sequential+block@bcdabcdabcdabcdabcdabcdabcdabcd1": Object {
              "complete": false,
              "description": null,
              "due": null,
              "effortActivities": 2,
              "effortTime": 15,
              "icon": null,
              "id": "block-v1:edX+DemoX+Demo_Course+type@sequential+block@bcdabcdabcdabcdabcdabcdabcdabcd1",
              "legacyWebUrl": "http://localhost:18000/courses/course-v1:edX+DemoX+Demo_Course/jump_to/block-v1:edX+DemoX+Demo_Course+type@sequential+block@bcdabcdabcdabcdabcdabcdabcdabcd1?experience=legacy",
              "sectionId": "block-v1:edX+DemoX+Demo_Course+type@chapter+block@bcdabcdabcdabcdabcdabcdabcdabcd2",
              "showLink": true,
              "title": "Title of Sequence",
            },
          },
        },
        "courseGoals": Object {
          "daysPerWeek": null,
          "goalOptions": Array [],
          "selectedGoal": null,
          "subscribedToReminders": null,
          "weeklyLearningGoalEnabled": false,
        },
        "courseTools": Array [
          Object {
            "analyticsId": "edx.bookmarks",
            "title": "Bookmarks",
            "url": "https://example.com/bookmarks",
          },
        ],
        "datesBannerInfo": Object {
          "contentTypeGatingEnabled": false,
          "missedDeadlines": false,
          "missedGatedContent": false,
        },
        "datesWidget": Object {
          "courseDateBlocks": Array [],
        },
        "enableProctoredExams": undefined,
        "enrollAlert": Object {
          "canEnroll": true,
          "extraText": "Contact the administrator.",
        },
        "enrollmentMode": undefined,
        "handoutsHtml": "<ul><li>Handout 1</li></ul>",
        "hasEnded": undefined,
        "hasScheduledContent": null,
        "id": "course-v1:edX+DemoX+Demo_Course",
        "offer": null,
        "resumeCourse": Object {
          "hasVisitedCourse": false,
          "url": "http://localhost:18000/courses/course-v1:edX+DemoX+Demo_Course/jump_to/block-v1:edX+Test+Block@12345abcde",
        },
        "timeOffsetMillis": 0,
        "userHasPassingGrade": undefined,
        "verifiedMode": Object {
          "accessExpirationDate": "2050-01-01T12:00:00",
          "currency": "USD",
          "currencySymbol": "$",
          "price": 149,
          "sku": "ABCD1234",
          "upgradeUrl": "http://localhost:18000/dashboard",
        },
        "welcomeMessageHtml": "<p>Welcome to this course!</p>",
      },
    },
  },
  "recommendations": Object {
    "recommendationsStatus": "loading",
  },
  "tours": Object {
    "showCoursewareTour": false,
    "showExistingUserCourseHomeTour": false,
    "showNewUserCourseHomeModal": false,
    "showNewUserCourseHomeTour": false,
    "toursEnabled": false,
  },
}
`;

exports[`Data layer integration tests Test fetchProgressTab Should fetch, normalize, and save metadata 1`] = `
Object {
  "courseHome": Object {
    "courseId": "course-v1:edX+DemoX+Demo_Course",
    "courseStatus": "loaded",
    "proctoringPanelStatus": "loading",
    "targetUserId": undefined,
    "toastBodyLink": null,
    "toastBodyText": null,
    "toastHeader": "",
  },
  "courseware": Object {
    "courseId": null,
    "courseStatus": "loading",
    "sequenceId": null,
    "sequenceMightBeUnit": false,
    "sequenceStatus": "loading",
  },
  "models": Object {
    "courseHomeMeta": Object {
      "course-v1:edX+DemoX+Demo_Course": Object {
        "canLoadCourseware": true,
        "celebrations": null,
        "courseAccess": Object {
          "additionalContextUserMessage": null,
          "developerMessage": null,
          "errorCode": null,
          "hasAccess": true,
          "userFragment": null,
          "userMessage": null,
        },
        "id": "course-v1:edX+DemoX+Demo_Course",
        "isEnrolled": false,
        "isMasquerading": false,
        "isSelfPaced": false,
        "isStaff": false,
        "number": "DemoX",
        "org": "edX",
        "originalUserIsStaff": false,
        "start": "2013-02-05T05:00:00Z",
        "tabs": Array [
          Object {
            "slug": "outline",
            "title": "Course",
            "url": "http://localhost:18000/courses/course-v1:edX+DemoX+Demo_Course/course/",
          },
          Object {
            "slug": "discussion",
            "title": "Discussion",
            "url": "http://localhost:18000/courses/course-v1:edX+DemoX+Demo_Course/discussion/forum/",
          },
          Object {
            "slug": "wiki",
            "title": "Wiki",
            "url": "http://localhost:18000/courses/course-v1:edX+DemoX+Demo_Course/course_wiki",
          },
          Object {
            "slug": "progress",
            "title": "Progress",
            "url": "http://localhost:18000/courses/course-v1:edX+DemoX+Demo_Course/progress",
          },
          Object {
            "slug": "instructor",
            "title": "Instructor",
            "url": "http://localhost:18000/courses/course-v1:edX+DemoX+Demo_Course/instructor",
          },
          Object {
            "slug": "dates",
            "title": "Dates",
            "url": "http://localhost:18000/courses/course-v1:edX+DemoX+Demo_Course/dates",
          },
        ],
        "title": "Demonstration Course",
        "userTimezone": "UTC",
        "username": "MockUser",
        "verifiedMode": Object {
          "accessExpirationDate": null,
          "currency": "USD",
          "currencySymbol": "$",
          "price": 149,
          "sku": "8CF08E5",
          "upgradeUrl": "http://localhost:18130/basket/add/?sku=8CF08E5",
        },
      },
    },
    "progress": Object {
      "course-v1:edX+DemoX+Demo_Course": Object {
        "accessExpiration": null,
        "certificateData": Object {},
        "completionSummary": Object {
          "completeCount": 1,
          "incompleteCount": 1,
          "lockedCount": 0,
        },
        "courseGrade": Object {
          "isPassing": true,
          "letterGrade": "pass",
          "percent": 1,
        },
        "courseId": "course-v1:edX+DemoX+Demo_Course",
        "creditCourseRequirements": null,
        "end": "3027-03-31T00:00:00Z",
        "enrollmentMode": "audit",
        "gradesFeatureIsFullyLocked": false,
        "gradesFeatureIsPartiallyLocked": false,
        "gradingPolicy": Object {
          "assignmentPolicies": Array [
            Object {
              "averageGrade": "1.00",
              "numDroppable": 1,
              "shortLabel": "HW",
              "type": "Homework",
              "weight": 1,
              "weightedGrade": 1,
            },
          ],
          "gradeRange": Object {
            "pass": 0.75,
          },
        },
        "hasScheduledContent": false,
        "id": "course-v1:edX+DemoX+Demo_Course",
        "sectionScores": Array [
          Object {
            "displayName": "First section",
            "subsections": Array [
              Object {
                "assignmentType": "Homework",
                "blockKey": "block-v1:edX+DemoX+Demo_Course+type@sequential+block@12345",
                "displayName": "First subsection",
                "hasGradedAssignment": true,
                "learnerHasAccess": true,
                "numPointsEarned": 0,
                "numPointsPossible": 3,
                "percentGraded": 0,
                "problemScores": Array [
                  Object {
                    "earned": 0,
                    "possible": 1,
                  },
                  Object {
                    "earned": 0,
                    "possible": 1,
                  },
                  Object {
                    "earned": 0,
                    "possible": 1,
                  },
                ],
                "showCorrectness": "always",
                "showGrades": true,
                "url": "http://learning.edx.org/course/course-v1:edX+Test+run/first_subsection",
              },
            ],
          },
          Object {
            "displayName": "Second section",
            "subsections": Array [
              Object {
                "assignmentType": "Homework",
                "displayName": "Second subsection",
                "hasGradedAssignment": true,
                "numPointsEarned": 1,
                "numPointsPossible": 1,
                "percentGraded": 1,
                "problemScores": Array [
                  Object {
                    "earned": 1,
                    "possible": 1,
                  },
                ],
                "showCorrectness": "always",
                "showGrades": true,
                "url": "http://learning.edx.org/course/course-v1:edX+Test+run/second_subsection",
              },
            ],
          },
        ],
        "studioUrl": "http://studio.edx.org/settings/grading/course-v1:edX+Test+run",
        "userHasPassingGrade": false,
        "verificationData": Object {
          "link": null,
          "status": "none",
          "statusDate": null,
        },
        "verifiedMode": null,
      },
    },
  },
  "recommendations": Object {
    "recommendationsStatus": "loading",
  },
  "tours": Object {
    "showCoursewareTour": false,
    "showExistingUserCourseHomeTour": false,
    "showNewUserCourseHomeModal": false,
    "showNewUserCourseHomeTour": false,
    "toursEnabled": false,
  },
}
`;
